network/src/dms.rs

```diff
@@ -359,109 +438,170 @@ impl<N: GossipNetwork, S: Storage> DistributedMessageSet<N, S> {
+            let result = async {
+                let message: RawMessage = serde_json::from_slice(&m)?;
+                let message = message.into_message()?;
+                this.read()
+                    .await
+                    .filter
+                    .filter(&message)
+                    .map_err(|e| anyhow!("{}", e))?;
+                Self::add_message_but_not_broadcast(
+                    &mut *this.read().await.storage.write().await,
+                    message,
+                )
+                .await?;
+                Result::<(), anyhow::Error>::Ok(())
             }
-            Result::<(), Error>::Ok(())
-        };
-        let storage = Arc::clone(&self.storage);
-        let rpc_task = async move {
-            run_server(
-                rpc_port,
-                [(
-                    "x".to_owned(),
-                    create_http_object(Arc::new(StorageWrapper { storage })
-                        as Arc<dyn DistributedMessageSetRpcInterface>),
-                )]
-                .iter()
-                .cloned()
-                .collect(),
-            )
             .await;
-            Ok(())
-        };
-        Ok(tokio::spawn(async move {
-            let x = try_join!(rpc_task, gossip_serve_task, broadcast_task, fetch_task);
-            match x {
-                Ok(_) => Ok(()),
-                Err(e) => Err(e),
+            if let Err(e) = result {
+                log::warn!("failed to receive a message from the gossip network: {}", e);
             }
-        }))
+        }
+        Ok(())
+    }
+
+    /// Serves the gossip network node and the RPC server indefinitely, constantly updating the storage.
+    pub async fn serve(
+        self,
+        network_config: NetworkConfig,
+        rpc_port: u16,
+        peers: SharedKnownPeers,
+    ) -> Result<Serve<Self, Error>, Error> {
+        let this = Arc::new(RwLock::new(self));
+        let rpc_task = tokio::spawn(Self::serve_rpc(Arc::clone(&this), rpc_port));
+        let fetch_task = tokio::spawn(Self::serve_fetch(
+            Arc::clone(&this),
+            network_config.clone(),
+            peers.clone(),
+        ));
+        let broadcast_task = tokio::spawn(Self::serve_broadcast(
+            Arc::clone(&this),
+            network_config.clone(),
+            peers.clone(),
+        ));
+        let gossip_task =
+            tokio::spawn(Self::serve_gossip(Arc::clone(&this), network_config, peers));
+        let (switch_send, switch_recv) = tokio::sync::oneshot::channel();
+        let this_ = Arc::clone(&this);
+        Ok(Serve::new(
+            tokio::spawn(async move {
+                let mut tasks =
+                    future::try_join4(rpc_task, fetch_task, broadcast_task, gossip_task).fuse();
+                let mut switch_recv = switch_recv.fuse();
+                select! {
+                    x = tasks => {
+                        let _ = switch_recv.await;
```

Comment:
Why do you await the termination signal here?

Reply:
I added comment

Reply:
Got it.